background/objectoriented.tex

```diff
@@ -1,210 +1,333 @@
+We look at the memory layout of a memory mapped Analoge-To-Digital and see how this can be represented as an object.
+
+\subsection{Memory Mapped I/O}
+Memory Mapped I/0, MMIO is a method for interfacing with peripheral devices in a computer system.
+The method entailes connecting the control registers of hardware devices to the same address bus as RAM.
+This results in a programming model where the programmer can use common memory operations to control the devices.
+
+Lets consider the Analoge-To-Digital, ADC, converter on the Gecko.
+The ADC converts an analoge signal to a digital representation.
+The ADC on the Gecko is memory mapped to the location \textbf{0x40002000}.
+This means that writing to a pointer pointing to this address will write to the control registers in the ADC device.
+A subsection of the ADC control register layout is given in \autoref{tab:back:memorymapped}.
 \begin{figure}[H]
   \centering
   \begin{tabular}{l|l|l|}
-    &&\\
-    \hline
     Location&Offset&Name\\
     \hline
+    &...&...\\
+    \hline
+    \hline
     0x40002000&0x0&CTRL\\
     \hline
     &0x4&CMD\\
     \hline
-    &0x8&STATUS\\
+    &...&...\\
+    \hline
+    &0xC&SINGLECTRL\\
     \hline
-    &...&... \\
+    &...&...\\
+    \hline
+    &0x24&SINGLEDATA\\
+    \hline
+    &...&...\\
     \hline
-    &0x3C&BIASPROG\\
     \hline
     &&\\
-
   \end{tabular}
-  \caption{ADC0 Memory map for EFM32GG}
+  \caption{Subsection of ADC0 Memory map for EFM32GG}
   \label{tab:back:memorymapped}
 \end{figure}
 
-\autoref{fig:back:objectmemory} shows the memory layout for the objects defined in \autoref{sec:back:oo}.
-We see that the difference between the Java managed language and the C and Rust system langauges is the tag field preciding the data fields for the object.
+\autoref{tab:back:memorymapped} shows the control register needed to perform a single ADC convertion.
+The ADC control register contains more register than in \autoref{tab:back:memorymapped}, butall registers in the ADC register file is aligned on a 4 byte boundry.
+The CTRL register is used to initialize the hardware device before performing a conversion, the CMD register is used to issue direct commands to the device like \emph{stop} and \emph{start}.
+We see that the CTRL register is at offset 0x0 from the base address of the ADC and that the CMD register is at offest 0x4 bytes.
+The two registers SINGLECTRL and SINGLEDATA are for initializing the single conversion and reading the results of a conversion repectivily.
+
+\subsection{Memory Layout of Objects}
+
+The traditional memory layout of an object in an object oriented language is an implementation detail.
+This is due to the fact that the fields of the object might have different sizes and optimizations can rearange the memory layout to optimize for size.
+Objects in a language like Java also includes a tag field at the base of the object as a reference to the class of the object in order to provide dynamic dispatch.
```

Comment:
Kan muligens nevne at data-layout i rust er udefinert pga dette.

Kan også nevne at man må bruke repr(c) for å være sikker på hvordan represntasjonen er, men kanskje det hører bedre til et annet sted